File: 10-逆向工程/02-移动端逆向/DEX文件格式解析/dex_format.go
// dex_format.go — DEX 文件格式的 Go 实现(与 dex_leb128.py / dex_format.py 同构)。
// 依据 AOSP《Dalvik 可执行文件格式》官方章节; 本机无 go 工具链: 人工审查 + 括号配平校验。
package main

import (
	"encoding/binary"
	"fmt"
)

// ---------- 常量(官方「位字段、字符串和常量定义」) ----------

var dexFileMagic = [8]byte{0x64, 0x65, 0x78, 0x0a, 0x30, 0x33, 0x39, 0x00} // "dex\n039\0"

const (
	endianConstant        = 0x12345678
	reverseEndianConstant = 0x78563412
	noIndex               = uint32(0xffffffff)
	headerSizeV40         = 0x70
	headerSizeV41         = 0x78
)

// map_list 类型代码(官方「类型代码」表)
const (
	typeHeaderItem             = 0x0000
	typeStringIDItem           = 0x0001
	typeTypeIDItem             = 0x0002
	typeProtoIDItem            = 0x0003
	typeFieldIDItem            = 0x0004
	typeMethodIDItem           = 0x0005
	typeClassDefItem           = 0x0006
	typeMapList                = 0x1000
	typeTypeList               = 0x1001
	typeClassDataItem          = 0x2000
	typeCodeItem               = 0x2001
	typeStringDataItem         = 0x2002
	typeDebugInfoItem          = 0x2003
	typeHiddenAPIDataItem      = 0xF000
)

// ---------- LEB128 ----------

// uleb128Encode: 除最后一字节外最高位均置 1, 未明确表示的位解译为 0
func uleb128Encode(v uint32) []byte {
	out := []byte{}
	for {
		b := byte(v & 0x7f)
		v >>= 7
		if v != 0 {
			out = append(out, b|0x80)
		} else {
			out = append(out, b)
			return out
		}
	}
}

// uleb128Decode 返回 (值, 消耗字节数)
func uleb128Decode(buf []byte) (uint32, int) {
	var result uint32
	var shift uint
	for i := 0; i < len(buf); i++ {
		b := buf[i]
		result |= uint32(b&0x7f) << shift
		shift += 7
		if b&0x80 == 0 {
			return result, i + 1
		}
	}
	return 0, len(buf)
}

// sleb128Encode: 最后一字节的最高载荷位做符号扩展
func sleb128Encode(v int32) []byte {
	out := []byte{}
	more := true
	for more {
		b := byte(uint32(v) & 0x7f)
		// Go 对 int32 的 >> 是算术右移, 天然符号扩展
		v >>= 7
		if (v == 0 && b&0x40 == 0) || (v == -1 && b&0x40 != 0) {
			more = false
		} else {
			b |= 0x80
		}
		out = append(out, b)
	}
	return out
}

func sleb128Decode(buf []byte) (int32, int) {
	var result uint32
	var shift uint
	var last byte
	n := 0
	for i := 0; i < len(buf); i++ {
		last = buf[i]
		result |= uint32(last&0x7f) << shift
		shift += 7
		n = i + 1
		if last&0x80 == 0 {
			break
		}
	}
	if last&0x40 != 0 {
		result -= uint32(1) << shift
	}
	return int32(result), n
}

// uleb128p1 = uleb128(v+1): 让 -1 编码成单字节, 且没有其他负数能编码
func uleb128p1Encode(v int32) []byte { return uleb128Encode(uint32(v) + 1) }

func uleb128p1Decode(buf []byte) (int32, int) {
	raw, n := uleb128Decode(buf)
	return int32(raw - 1), n
}

// ---------- MUTF-8 ----------

// utf16Units: 增补平面字符拆成两个 UTF-16 码元
func utf16Units(cp uint32) []uint16 {
	if cp < 0x10000 {
		return []uint16{uint16(cp)}
	}
	v := cp - 0x10000
	return []uint16{uint16(0xD800 + (v >> 10)), uint16(0xDC00 + (v & 0x3FF))}
}

// utf16Size = 官方 string_data_item.utf16_size: 以 UTF-16 代码单元计, 不是字节数
func utf16Size(s string) int {
	n := 0
	for _, r := range s {
		n += len(utf16Units(uint32(r)))
	}
	return n
}

// mutf8Encode: U+0000 编成 0xc0 0x80, 结尾补一个 0 字节
func mutf8Encode(s string) []byte {
	out := []byte{}
	for _, r := range s {
		for _, u := range utf16Units(uint32(r)) {
			switch {
			case u == 0x0000:
				out = append(out, 0xc0, 0x80)
			case u <= 0x7f:
				out = append(out, byte(u))
			case u <= 0x7ff:
				out = append(out, byte(0xc0|(u>>6)), byte(0x80|(u&0x3f)))
			default:
				out = append(out, byte(0xe0|(u>>12)), byte(0x80|((u>>6)&0x3f)), byte(0x80|(u&0x3f)))
			}
		}
	}
	return append(out, 0x00)
}

// mutf8Decode 还原码元序列(不做排序语义)
func mutf8Decode(buf []byte) []uint16 {
	units := []uint16{}
	for i := 0; i < len(buf) && buf[i] != 0x00; {
		b0 := buf[i]
		switch {
		case b0 < 0x80:
			units = append(units, uint16(b0))
			i++
		case b0&0xe0 == 0xc0:
			units = append(units, uint16(b0&0x1f)<<6|uint16(buf[i+1]&0x3f))
			i += 2
		case b0&0xf0 == 0xe0:
			units = append(units, uint16(b0&0x0f)<<12|uint16(buf[i+1]&0x3f)<<6|uint16(buf[i+2]&0x3f))
			i += 3
		default:
			panic("非法 MUTF-8 起始字节")
		}
	}
	return units
}

// ---------- header_item ----------

type dexHeader struct {
	magic        [8]byte
	checksum     uint32
	signature    [20]byte
	fileSize     uint32
	headerSize   uint32
	endianTag    uint32
	linkSize     uint32
	linkOff      uint32
	mapOff       uint32
	stringIDsSize uint32
	stringIDsOff uint32
	typeIDsSize  uint32
	typeIDsOff   uint32
	protoIDsSize uint32
	protoIDsOff  uint32
	fieldIDsSize uint32
	fieldIDsOff  uint32
	methodIDsSize uint32
	methodIDsOff uint32
	classDefsSize uint32
	classDefsOff uint32
	dataSize     uint32
	dataOff      uint32
}

// parseHeader 按官方 header_item 逐字段解析(小端, v40 版 0x70 字节)
func parseHeader(buf []byte) (*dexHeader, error) {
	if len(buf) < headerSizeV40 {
		return nil, fmt.Errorf("文件短于 header_item 的 0x70 字节")
	}
	h := &dexHeader{}
	copy(h.magic[:], buf[0:8])
	for i, b := range dexFileMagic {
		if h.magic[i] != b {
			return nil, fmt.Errorf("magic 不是 dex\\n039\\0")
		}
	}
	h.checksum = binary.LittleEndian.Uint32(buf[8:12])
	copy(h.signature[:], buf[12:32])
	h.fileSize = binary.LittleEndian.Uint32(buf[32:36])
	h.headerSize = binary.LittleEndian.Uint32(buf[36:40])
	h.endianTag = binary.LittleEndian.Uint32(buf[40:44])
	if h.endianTag != endianConstant && h.endianTag != reverseEndianConstant {
		return nil, fmt.Errorf("endian_tag 0x%08x 既不是 ENDIAN_CONSTANT 也不是 REVERSE", h.endianTag)
	}
	h.linkSize = binary.LittleEndian.Uint32(buf[44:48])
	h.linkOff = binary.LittleEndian.Uint32(buf[48:52])
	h.mapOff = binary.LittleEndian.Uint32(buf[52:56])
	h.stringIDsSize = binary.LittleEndian.Uint32(buf[56:60])
	h.stringIDsOff = binary.LittleEndian.Uint32(buf[60:64])
	h.typeIDsSize = binary.LittleEndian.Uint32(buf[64:68])
	h.typeIDsOff = binary.LittleEndian.Uint32(buf[68:72])
	h.protoIDsSize = binary.LittleEndian.Uint32(buf[72:76])
	h.protoIDsOff = binary.LittleEndian.Uint32(buf[76:80])
	h.fieldIDsSize = binary.LittleEndian.Uint32(buf[80:84])
	h.fieldIDsOff = binary.LittleEndian.Uint32(buf[84:88])
	h.methodIDsSize = binary.LittleEndian.Uint32(buf[88:92])
	h.methodIDsOff = binary.LittleEndian.Uint32(buf[92:96])
	h.classDefsSize = binary.LittleEndian.Uint32(buf[96:100])
	h.classDefsOff = binary.LittleEndian.Uint32(buf[100:104])
	h.dataSize = binary.LittleEndian.Uint32(buf[104:108])
	h.dataOff = binary.LittleEndian.Uint32(buf[108:112])
	return h, nil
}

// codeItemPadding: 官方 —— 只有 tries_size 非零**且** insns_size 是奇数时才存在 2 字节填充
func codeItemPadding(triesSize, insnsSize int) int {
	if triesSize != 0 && insnsSize%2 == 1 {
		return 1
	}
	return 0
}

func codeItemUnits(triesSize, insnsSize int) int {
	return 8 + insnsSize + codeItemPadding(triesSize, insnsSize)
}

File: 10-逆向工程/02-移动端逆向/DEX文件格式解析/dex_format.py
"""DEX 文件格式：按 AOSP 官方《Dalvik 可执行文件格式》实现的编码/解码与最小文件构造。

结构对照（官方文档章节）：
  LEB128 / uleb128p1  -> "LEB128"
  header_item         -> "项和相关结构 / header_item"
  string_id_item      -> string_ids 区段
  string_data_item    -> data 区段（MUTF-8）
  proto_id_item / type_list / class_def_item / class_data_item / code_item
参考：https://source.android.google.cn/docs/core/dalvik/dex-format
"""

import struct

from dex_leb128 import uleb128_decode, mutf8_decode, mutf8_to_str

# ---------- 常量（官方文档「位字段、字符串和常量定义」） ----------

DEX_FILE_MAGIC = b"dex\n039\x00"          # 039 版：0x64 0x65 0x78 0x0a 0x30 0x33 0x39 0x00
ENDIAN_CONSTANT = 0x12345678
REVERSE_ENDIAN_CONSTANT = 0x78563412
NO_INDEX = 0xFFFFFFFF                      # uint NO_INDEX == -1（按有符号解读）

HEADER_SIZE_V40 = 0x70                     # v40 及以下：112 字节
HEADER_SIZE_V41 = 0x78                     # v41 及以上：120 字节（新增 container_size/header_offset）

# map_list 类型代码（官方「类型代码」表）
TYPE_HEADER_ITEM = 0x0000
TYPE_STRING_ID_ITEM = 0x0001
TYPE_TYPE_ID_ITEM = 0x0002
TYPE_PROTO_ID_ITEM = 0x0003
TYPE_FIELD_ID_ITEM = 0x0004
TYPE_METHOD_ID_ITEM = 0x0005
TYPE_CLASS_DEF_ITEM = 0x0006
TYPE_CALL_SITE_ID_ITEM = 0x0007
TYPE_METHOD_HANDLE_ITEM = 0x0008
TYPE_MAP_LIST = 0x1000
TYPE_TYPE_LIST = 0x1001
TYPE_ANNOTATION_SET_REF_LIST = 0x1002
TYPE_ANNOTATION_SET_ITEM = 0x1003
TYPE_CLASS_DATA_ITEM = 0x2000
TYPE_CODE_ITEM = 0x2001
TYPE_STRING_DATA_ITEM = 0x2002
TYPE_DEBUG_INFO_ITEM = 0x2003
TYPE_ANNOTATION_ITEM = 0x2004
TYPE_ENCODED_ARRAY_ITEM = 0x2005
TYPE_ANNOTATIONS_DIRECTORY_ITEM = 0x2006
TYPE_HIDDENAPI_CLASS_DATA_ITEM = 0xF000

# 定长项的大小（官方类型代码表最后一列）
FIXED_ITEM_SIZE = {
    TYPE_HEADER_ITEM: 0x70,
    TYPE_STRING_ID_ITEM: 0x04,
    TYPE_TYPE_ID_ITEM: 0x04,
    TYPE_PROTO_ID_ITEM: 0x0C,
    TYPE_FIELD_ID_ITEM: 0x08,
    TYPE_METHOD_ID_ITEM: 0x08,
    TYPE_CLASS_DEF_ITEM: 0x20,
    TYPE_CALL_SITE_ID_ITEM: 0x04,
    TYPE_METHOD_HANDLE_ITEM: 0x08,
}



def code_item_padding(tries_size, insns_size):
    """官方：padding 只有在 tries_size 非零**且** insns_size 是奇数时才存在（2 字节）。"""
    return 1 if (tries_size != 0 and insns_size % 2 == 1) else 0


def code_item_units(tries_size, insns_size):
    """code_item 除 insns 之外占用的 16 位代码单元数（含可选 padding）。"""
    return 8 + insns_size + code_item_padding(tries_size, insns_size)


# ---------- 解析 ----------

class DexFile(object):

    def __init__(self, blob):
        self.blob = blob
        self._parse_header()
        self._parse_map()

    def _u32(self, off):
        return struct.unpack_from("<I", self.blob, off)[0]

    def _parse_header(self):
        assert self.blob[0:8] == DEX_FILE_MAGIC, "magic 不是 dex\\n039\\0"
        self.checksum = self._u32(8)
        self.signature = self.blob[12:32]
        self.file_size = self._u32(32)
        self.header_size = self._u32(36)
        self.endian_tag = self._u32(40)
        assert self.endian_tag in (ENDIAN_CONSTANT, REVERSE_ENDIAN_CONSTANT)
        self.map_off = self._u32(52)
        self.string_ids_size = self._u32(56)
        self.string_ids_off = self._u32(60)
        self.type_ids_size = self._u32(64)
        self.type_ids_off = self._u32(68)
        self.proto_ids_size = self._u32(72)
        self.proto_ids_off = self._u32(76)
        self.field_ids_size = self._u32(80)
        self.method_ids_size = self._u32(88)
        self.method_ids_off = self._u32(92)
        self.class_defs_size = self._u32(96)
        self.class_defs_off = self._u32(100)
        self.data_size = self._u32(104)
        self.data_off = self._u32(108)

    def _parse_map(self):
        size = self._u32(self.map_off)
        self.map = []
        for i in range(size):
            off = self.map_off + 4 + i * 12
            t, _unused, count, item_off = struct.unpack_from("<HHII", self.blob, off)
            self.map.append({"type": t, "size": count, "offset": item_off})

    # --- 字符串 ---
    def string_at(self, i):
        data_off = self._u32(self.string_ids_off + 4 * i)
        utf16_len, pos = uleb128_decode(self.blob, data_off)
        raw = bytearray()
        while self.blob[pos] != 0x00:
            raw.append(self.blob[pos])
            pos += 1
        assert utf16_len == len(mutf8_decode(bytes(raw))), "utf16_size 与解码长度不符"
        return mutf8_to_str(bytes(raw)), utf16_len

    def type_at(self, i):
        return self.string_at(self._u32(self.type_ids_off + 4 * i))[0]

    def proto_at(self, i):
        base = self.proto_ids_off + 12 * i
        shorty, ret, params_off = struct.unpack_from("<III", self.blob, base)
        params = []
        if params_off:
            n = self._u32(params_off)
            for k in range(n):
                params.append(self._u32(params_off + 4) if False else
                              struct.unpack_from("<H", self.blob, params_off + 4 + 2 * k)[0])
        return {"shorty": self.string_at(shorty)[0],
                "return_type": self.type_at(ret),
                "params": [self.type_at(t) for t in params]}

    def method_at(self, i):
        base = self.method_ids_off + 8 * i
        cls, name, proto = struct.unpack_from("<HHI", self.blob, base)
        return {"class": self.type_at(cls), "name": self.string_at(name)[0],
                "proto": self.proto_at(proto)}

    def class_def_at(self, i):
        base = self.class_defs_off + 32 * i
        (cls, flags, sup, ifaces, src, anno, cdata, svals) = struct.unpack_from("<IIIIIIII", self.blob, base)
        return {"class": self.type_at(cls), "access_flags": flags,
                "superclass": NO_INDEX if sup == NO_INDEX else self.type_at(sup),
                "interfaces_off": ifaces, "source_file_idx": src,
                "annotations_off": anno, "class_data_off": cdata,
                "static_values_off": svals}

    def class_data_at(self, off):
        pos = off
        static_n, pos = uleb128_decode(self.blob, pos)
        inst_n, pos = uleb128_decode(self.blob, pos)
        direct_n, pos = uleb128_decode(self.blob, pos)
        virtual_n, pos = uleb128_decode(self.blob, pos)
        methods = []
        prev = 0
        for _ in range(direct_n):
            diff, pos = uleb128_decode(self.blob, pos)
            flags, pos = uleb128_decode(self.blob, pos)
            code_off, pos = uleb128_decode(self.blob, pos)
            prev += diff                    # method_idx_diff 是相对前一个元素的差值
            methods.append({"method_idx": prev, "access_flags": flags, "code_off": code_off})
        return {"static_fields_size": static_n, "instance_fields_size": inst_n,
                "direct_methods_size": direct_n, "virtual_methods_size": virtual_n,
                "direct_methods": methods}

    def code_at(self, off):
        (regs, ins, outs, tries) = struct.unpack_from("<HHHH", self.blob, off)
        debug = self._u32(off + 8)
        insns_size = self._u32(off + 12)
        pos = off + 16
        insns = [struct.unpack_from("<H", self.blob, pos + 2 * k)[0] for k in range(insns_size)]
        return {"registers_size": regs, "ins_size": ins, "outs_size": outs,
                "tries_size": tries, "debug_info_off": debug,
                "insns_size": insns_size, "insns": insns}
